src/map/pc.c

```diff
@@ -1624,33 +1624,30 @@ static int pc_calc_skilltree(struct map_session_data *sd)
 			sd->status.skill[i].lv = (sd->status.skill[i].flag == SKILL_FLAG_TEMPORARY) ? 0 : sd->status.skill[i].flag - SKILL_FLAG_REPLACED_LV_0;
 			sd->status.skill[i].flag = SKILL_FLAG_PERMANENT;
 		}
+	}
 
-		if (sd->sc.count && sd->sc.data[SC_SOULLINK] && sd->sc.data[SC_SOULLINK]->val2 == SL_BARDDANCER
-		 && ((skill->dbs->db[i].nameid >= BA_WHISTLE && skill->dbs->db[i].nameid <= BA_APPLEIDUN)
-		  || (skill->dbs->db[i].nameid >= DC_HUMMING && skill->dbs->db[i].nameid <= DC_SERVICEFORYOU))
-		) {
-			//Enable Bard/Dancer spirit linked skills.
-			int linked_nameid = skill->get_linked_song_dance_id(skill->dbs->db[i].nameid);
-			if (linked_nameid == 0) {
-				Assert_report("Linked bard/dance skill not found");
-				continue;
+	//Enable Bard/Dancer spirit linked skills.
+	if (sd->sc.count && sd->sc.data[SC_SOULLINK] && sd->sc.data[SC_SOULLINK]->val2 == SL_BARDDANCER) {
+		for (int i = 0; i < MAX_SKILL_DB; i++) {
```

Comment:
You definitely shouldn't be looping over all skills again. Why was it necessary to move this out of the upper loop? Instead, can't a change of order help here if the loops first if statement is the initial problem?

Reply:
>You definitely shouldn't be looping over all skills again

that's because hercules has compress the skill ID into skill index, which Ind made to save server memory
https://herc.ws/board/topic/512-skill-id-processing-overhaul/
if let's say for example, I add a custom skill with ID ...
```
static const struct {
	int start;
	int end;
} skill_idx_ranges[] = {
	{ NV_BASIC, NPC_LEX_AETERNA },
// somewhere here
	{ KN_CHARGEATK, SA_ELEMENTWIND },
	{ RK_ENCHANTBLADE, AB_SILENTIUM },
	{ WL_WHITEIMPRISON, SC_FEINTBOMB },
	{ LG_CANNONSPEAR, SR_GENTLETOUCH_REVITALIZE },
	{ WA_SWING_DANCE, WA_MOONLIT_SERENADE },
	{ MI_RUSH_WINDMILL, MI_HARMONIZE },
	{ WM_LESSON, WM_UNLIMITED_HUMMING_VOICE },
	{ SO_FIREWALK, SO_EARTH_INSIGNIA },
	{ GN_TRAINING_SWORD, GN_SLINGITEM_RANGEMELEEATK },
	{ AB_SECRAMENT, LG_OVERBRAND_PLUSATK },
	{ ALL_ODINS_RECALL, ALL_LIGHTGUARD },
	{ RL_GLITTERING_GREED, RL_GLITTERING_GREED_ATK },
	{ KO_YAMIKUMO, OB_AKAITSUKI },
	{ ECL_SNOWFLIP, ALL_THANATOS_RECALL },
	{ GC_DARKCROW, NC_MAGMA_ERUPTION_DOTDAMAGE },
	{ SU_BASIC_SKILL, SU_SPIRITOFSEA },
	{ HLIF_HEAL, MH_VOLCANIC_ASH },
	{ MS_BASH, MER_INVINCIBLEOFF2 },
	{ EL_CIRCLE_OF_FIRE, EL_STONE_RAIN },
	{ GD_APPROVAL, GD_DEVELOPMENT },
```
so please **DO NOT assume the skill index will stay the same for all server**
some server has already made their own custom skill ID

I think that's the reason why Haru made that commit in the first place https://github.com/HerculesWS/Hercules/commit/bfa423fea55aa5659506cf85e6c418346804ac89

------------------------------------------------

> Instead, can't a change of order help here if the loops first if statement is the initial problem?

with clean hercules with no custom skills, bard skill index is higher than dancer

do you have any better idea how to do it then ?

Reply:
EDIT -> I think I get it now
use ARRAY
ok I make another one

Reply:
nice solution, good thinking.